UnwrapBlog.md

```diff
@@ -0,0 +1,92 @@
+
+```ts
+function unwrap(input: Input<...MaybeHasInputsOrPromisesOrOutputsArbitrarilyDeeper...>): SomeThingWithAllTheInputsPromisesAndOutputsErased;
+```
+
+This was a big headscratcher for a while.  But after exploring some of the powerful new features that TypeScript has been bringing about in recent releases, we finally found a great way to manage it.  In the end, it ends up looking something like this:
+
+```ts
+type primitive = string | number | boolean | undefined | null;
+
+type Unwrap<T> =
+    T extends Promise<infer U> ? UnwrapSimple<U> :
+    T extends Output<infer U> ? UnwrapSimple<U> :
+    UnwrapSimple<T>;
+
+type UnwrapSimple<T> =
+    T extends primitive ? T :
+    T extends Array<infer U> ? UnwrappedArray<U> :
+    T extends object ? UnwrappedObject<T> :
+    never;
+
+interface UnwrappedArray<T> extends Array<Unwrap<T>> {}
+
+type UnwrappedObject<T> = {
+    [P in keyof T]: Unwrap<T[P]>;
+};
+
+function unwrap<T>(val: T): Unwrap<T>;
+```
+
+It's a bit of a doozy containing a bunch of type definitions and the final function declaration that uses them.  It makes use of many advanced TypeScript typing features.  Including 'Union Types' (the 'primitive' type), 'Mapped Types' (which 'UnwrappedObject' is an example of), 'Conditional Types' (the types with `?` and `:` in them), 'Inferred Types' (the use of `infer U` in several places).  You can read more about these in the [Advanced Types](https://www.typescriptlang.org/docs/handbook/advanced-types.html) documentation.
+
+Put together, the `Unwrap<T>` is effectively itself a recursively defined structural pattern matching type.  You can see it reference itself once you walk into `UnwrappedArray<T>` and `UnwrappedObject<T>` types. As it recurses, it matches the type it has and then 'peels away' the wrappings that have been caused by `Promise<T>` and `Output<T>`.  As it hits Arrays and Objects, it dives deeper, unwrapping along the way.  This is all manageable in TypeScript because it does this sort of matching and unwrapping in a lazy fashion, allowing us to express what is effectively an infinitely large pattern-matching type, without it having any trouble understanding it or applying it!
+
+It amazes us how well this works and how amazing TypeScript is here.  As you can see here, TS completely understands what's going on and will give you all the great type-checking support and tooling that we know and love:
+
+![image](https://user-images.githubusercontent.com/4564579/45712608-9e6cd100-bb41-11e8-90dd-b4fe79d236fe.png)
+
+Since adding support for this, we've been able to start using this new function across our own codebase and we've seen things get dramatically simpler.  The pyramid-of-doom goes away and code becomes much cleaner and easier to read and maintain.  JavaScript helped us write this system easily, but thanks to TypeScript we can express these complex intents and have the safety and understanding to do this sort of thing with confidence.
```

Comment:
Any good before/after example to show?

Reply:
I didn't want to point at actual examples, because the blog post is a bit of a simplification.  For example, our 'unwrap' is actually called 'output' (and i didn't want to explain that), and we return an `Output<Unwrap<T>>` and i didn't want to have to explain that.  And the code that uses that needs to use .apply, and i didn't want to explain that :)

The meat of the type system and how we use it is what i was trying to convey here.  And, to keep things on that point, i simplified things i felt would just not substantively contribute to that message.